forest/cli/src/cli/snapshot_cmd.rs

```diff
@@ -323,6 +356,129 @@ fn clean(config: &Config, snapshot_dir: &Option<PathBuf>, force: bool) -> anyhow
     Ok(())
 }
 
+async fn validate(
+    config: &Config,
+    recent_stateroots: &i64,
+    snapshot: &PathBuf,
+    force: bool,
+) -> anyhow::Result<()> {
+    let confirm = force
+        || atty::is(atty::Stream::Stdin)
+            && Confirm::with_theme(&ColorfulTheme::default())
+                .with_prompt(format!(
+                    "This will result in using approximately {} MB of data. Proceed?",
+                    std::fs::metadata(snapshot)?.len() / (1024 * 1024)
+                ))
+                .default(false)
+                .interact()
+                .unwrap_or_default();
+
+    if confirm {
+        let tmp_db_path = TempDir::new()?;
+
+        let db_path = tmp_db_path.path().join(&config.chain.name);
+
+        let db = open_db(
+            &db_path,
+            #[cfg(feature = "rocksdb")]
+            config,
+        )?;
+
+        let chain_store = Arc::new(ChainStore::new(db).await);
+
+        let genesis = read_genesis_header(
+            config.client.genesis_file.as_ref(),
+            config.chain.genesis_bytes(),
+            &chain_store,
+        )
+        .await?;
+
+        let cids = {
+            let file = tokio::fs::File::open(&snapshot).await?;
+            let reader = FetchProgress::fetch_from_file(file).await?;
+            load_car(chain_store.blockstore(), reader.compat()).await?
+        };
+
+        let ts = chain_store.tipset_from_keys(&TipsetKeys::new(cids)).await?;
+
+        validate_links_and_genesis_traversal(
+            &chain_store,
+            ts,
+            chain_store.blockstore(),
+            *recent_stateroots,
+            &genesis,
+            &config.chain.name,
+        )
+        .await?;
+    }
+
+    Ok(())
+}
+
+async fn validate_links_and_genesis_traversal<DB>(
+    chain_store: &ChainStore<DB>,
+    ts: Arc<Tipset>,
+    db: &DB,
+    recent_stateroots: ChainEpoch,
+    genesis_tipset: &Tipset,
+    network: &str,
+) -> anyhow::Result<()>
+where
+    DB: fvm_ipld_blockstore::Blockstore + Store + Send + Sync,
+{
+    let mut seen = std::collections::HashSet::<Cid>::new();
+    let upto = ts.epoch() - recent_stateroots;
+
+    let mut tsk = ts.parents().clone();
+
+    let total_size = ts.epoch();
+    let pb = forest_utils::io::ProgressBar::new(total_size as u64);
+    pb.message("Validating tipsets: ");
+    pb.set_max_refresh_rate(Some(std::time::Duration::from_millis(500)));
+
+    // Security: Recursive snapshots are difficult to create but not impossible. This
+    // limits the amount of recursion we do.
+    let mut prev_epoch = ts.epoch();
+    loop {
```

Comment:
For readability, it may be worth putting the content of this loop in a separate method. Then, potentially, we could do some basic unit testing.